packages/liveblocks-react/src/comments/lib/revalidation.ts

```diff
@@ -0,0 +1,364 @@
+        const activeMutation = manager.mutation;
+        if (
+          activeMutation &&
+          beforeMutationTimestamp !== activeMutation.startTime
+        ) {
+          return;
+        }
+      } catch (err) {
+        // If the mutation request fails, revert the optimistic update and throw the error
+        manager.cache = currentCache;
+        throw err;
+      }
+
+      // Mark the mutation as completed by setting the end time to the current timestamp
+      manager.mutation = {
+        startTime: beforeMutationTimestamp,
+        endTime: ++timestamp,
+      };
+
+      // Deleting the concurrent request markers so new requests will not be deduped.
+      manager.request = undefined;
+      void revalidateCache(false);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps -- ESLint recommends against adding `manager` as a dependency, but not doing so causes the code inside `useEffect` to reference an outdated version of `manager`
+    [manager, revalidateCache]
+  );
+
+  return mutate;
+}
+
+/**
+ * This hook automatically revalidates the cache when the browser comes back online, when the document becomes visible, or periodically.
+ * @param manager - The cache manager
+ * @param revalidateCache - The function used to revalidate the cache
+ * @param options.revalidateOnFocus - If true, the cache will be revalidated when the document becomes visible
+ * @param options.revalidateOnReconnect - If true, the cache will be revalidated when the browser comes back online
+ * @param options.refreshInterval - The interval (in milliseconds) at which the cache should be revalidated
+ */
+export function useAutomaticRevalidation<Data>(
+  manager: CacheManager<Data>,
+  revalidateCache: (shouldDedupe: boolean) => Promise<void>,
+  options: {
+    revalidateOnFocus?: boolean;
+    revalidateOnReconnect?: boolean;
+    refreshInterval?: number;
+  } = {}
+) {
+  const isOnlineRef = useRef(true); // Stores the current online status of the browser
+
+  const {
+    revalidateOnFocus = true,
+    revalidateOnReconnect = true,
+    refreshInterval = 0,
+  } = options;
+
+  /**
+   * Periodically revalidate the cache. The revalidation is skipped if the browser is offline or if the document is not visible.
+   */
+  useEffect(() => {
+    let revalidationTimerId: number;
+
+    function scheduleRevalidation() {
+      if (refreshInterval === 0) return;
+
+      revalidationTimerId = window.setTimeout(() => {
+        const isOnline = isOnlineRef.current;
+        const isVisible = document.visibilityState === "visible";
+
+        // Only revalidate if the browser is online AND document is visible AND there are currently no errors, otherwise schedule the next revalidation
+        if (isOnline && isVisible && !manager.cache?.error) {
+          // Revalidate cache and then schedule the next revalidation
+          void revalidateCache(true).then(scheduleRevalidation);
```

Comment:
We only revalidate if the browser is online **AND** document is visible **AND** there are currently no errors, otherwise we schedule the next revalidation.